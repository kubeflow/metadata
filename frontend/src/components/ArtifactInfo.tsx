/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import {stylesheet} from 'typestyle';
import {color, commonCss} from '../Css';
import {formatDateString, getArtifactProperty} from '../lib/Utils';
import {MlMetadataArtifact} from '../apis/service';
import {ArtifactProperties, ArtifactCustomProperties} from '../lib/Api';
import {ModelInfo} from './ModelInfo';

export const css = stylesheet({
  artifactInfo: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  field: {
    flexBasis: '300px',
    marginBottom: '32px',
  },
  term: {
    color: color.grey,
    fontSize: '12px',
    letterSpacing: '0.2px',
    lineHeight: '16px',
  },
  value: {
    color: color.secondaryText,
    fontSize: '14px',
    letterSpacing: '0.2px',
    lineHeight: '20px',
  }
});

export interface ArtifactInfoProps {
  typeName: string;
  artifact: MlMetadataArtifact;
}

export class ArtifactInfo extends React.Component<ArtifactInfoProps, {}> {

  public render(): JSX.Element {
    const {artifact} = this.props;
    const createTime = formatDateString(
      getArtifactProperty(artifact, ArtifactProperties.CREATE_TIME) || '');
    return (
      <section>
        <h1 className={commonCss.header}>{this.props.typeName} info</h1>
        <dl className={css.artifactInfo}>
          <div className={css.field}>
            <dt className={css.term}>Version ID</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactProperties.VERSION)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Workspace</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactCustomProperties.WORKSPACE,
                true)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Run</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactCustomProperties.RUN,
                true)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Description</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactProperties.DESCRIPTION)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Creation time</dt>
            <dd className={css.value}>{createTime}</dd>
          </div>
          {this.props.typeName === 'Model' ?
            <ModelInfo model={artifact} /> :
            <div className={css.field}>
              <dt className={css.term}>URI</dt>
              <dd className={css.value}>{artifact.uri}</dd>
            </div>
          }
        </dl>
      </section>
    );
  }
}
