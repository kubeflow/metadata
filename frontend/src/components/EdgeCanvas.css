.edgeCanvas {
  position: absolute;
  width: 160px;
  height: 440px;
  border: 0;
  display: inline-block;
  margin-left: 260px;
  margin-top: 74px;
  z-index: -1;
}
.edgeCanvas.reverse {
  margin-left: 0;
  transform: translateX(-100%);
}

.kWKfgJ {
  display: block;
  overflow: visible;
  position: absolute;
}