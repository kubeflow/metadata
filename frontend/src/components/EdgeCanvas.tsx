import * as React from 'react';
import LineChart from 'react-svg-line-chart';
import {classes, stylesheet} from 'typestyle';
import {LineageCardType} from './LineageTypes';
import './LineChart.d.ts';
import {px} from '../Css';
import {CARD_WIDTH, EDGE_WIDTH} from "./LineageCss";

interface EdgeCanvasProps {
  cardArray: number[];
  reverseEdges: boolean;
  type: LineageCardType;
}

export const EdgeCanvas: React.FC<EdgeCanvasProps> = ({cardArray, reverseEdges}) => {
  let viewHeight = 1;

  const cardBodyHeight = 67;
  const cardContainerSpacerHeight = 24;
  const cardTitleHeight = 41;
  const cardTitleBorders = 2;
  const cardOffset = cardContainerSpacerHeight + cardTitleHeight + cardTitleBorders;

  const css = stylesheet({
    edgeCanvas: {
      border: 0,
      display: 'block',
      height: px(440),
      marginLeft: px(CARD_WIDTH),
      marginTop: px(74),
      overflow: 'visible',
      position: 'absolute',
      width: EDGE_WIDTH,
      zIndex: -1,
      '$nest': {
        svg: {
          display: 'block',
          overflow: 'visible',
          position: 'absolute',
        }
      }
    },
    edgeCanvasReverse: {
      marginLeft: 0,
      transform: 'translateX(-100%)',
    },
  });

  const lastNode = reverseEdges ? 'y1' : 'y4';
  const lastNodePositions = {
    y1: 0,
    y4: 0,
  };

  const edgeLines: JSX.Element[] = [];
  cardArray.forEach((rows, i) => {
    for (let j = 0; j < rows; j++) {
      const {y1, y4} = lastNodePositions;
      edgeLines.push(
        <LineChart
          key={`${i}-${j}`}
          data={[
            {x: 0, y: y1},
            {x: 30, y: y1},
            {x: EDGE_WIDTH - 30, y: y4},
            {x: EDGE_WIDTH, y: y4},
          ]}
          areaVisible={false}
          axisVisible={false}
          gridVisible={false}
          labelsVisible={false}
          pathColor={'#BDC1C6'}
          pathVisible={true}
          pathWidth={1}
          pathOpacity={1}
          pointsVisible={false}
          viewBoxHeight={viewHeight}
          viewBoxWidth={EDGE_WIDTH}
          pathSmoothing={0}
        />
      );
      viewHeight += cardBodyHeight;
      lastNodePositions[lastNode] += cardBodyHeight;
    }
    viewHeight += cardOffset;
    lastNodePositions[lastNode] += cardOffset;
  });

  const edgeCanvasClasses = classes(css.edgeCanvas, reverseEdges ? css.edgeCanvasReverse: '');
  return  <div className={edgeCanvasClasses}>{edgeLines}</div>
};
