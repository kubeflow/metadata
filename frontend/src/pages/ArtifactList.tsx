/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import CustomTable, {Column, Row, ExpandState, CustomRendererProps} from '../components/CustomTable';
import {Page} from './Page';
import {ToolbarProps} from '../components/Toolbar';
import {classes} from 'typestyle';
import {commonCss, padding} from '../Css';
import {getResourceProperty, rowCompareFn, rowFilterFn, groupRows, getExpandedRow} from '../lib/Utils';
import {Api, ArtifactProperties, ArtifactCustomProperties, ListRequest} from '../lib/Api';
import {MlMetadataArtifact} from '../apis/service/api';
import {RoutePage, RouteParams} from '../components/Router';
import {Link} from 'react-router-dom';
import {ArtifactType} from "../generated/src/apis/metadata/metadata_store_pb";
import {GetArtifactTypesRequest} from "../generated/src/apis/metadata/metadata_store_service_pb";

interface ArtifactListState {
  artifacts: MlMetadataArtifact[];
  rows: Row[];
  expandedRows: Map<number, Row[]>;
  columns: Column[];
}

class ArtifactList extends Page<{}, ArtifactListState> {
  private tableRef = React.createRef<CustomTable>();
  private api = Api.getInstance();
  private artifactTypes: Map<number, ArtifactType>;
  private nameCustomRenderer: React.FC<CustomRendererProps<string>> =
    (props: CustomRendererProps<string>) => {
      const [artifactType, artifactId] = props.id.split(':');
      const link = RoutePage.ARTIFACT_DETAILS
        .replace(`:${RouteParams.ARTIFACT_TYPE}+`, artifactType)
        .replace(`:${RouteParams.ID}`, artifactId);
      return (
        <Link onClick={(e) => e.stopPropagation()}
          className={commonCss.link}
          to={link}>
          {props.value}
        </Link>
      );
  }

  constructor(props: any) {
    super(props);
    this.state = {
      artifacts: [],
      columns: [
        {
          label: 'Pipeline/Workspace',
          flex: 2,
          customRenderer: this.nameCustomRenderer,
          sortKey: 'pipelineName'
        },
        {
          label: 'Name',
          flex: 1,
          customRenderer: this.nameCustomRenderer,
          sortKey: 'name',
        },
        {label: 'ID', flex: 1, sortKey: 'id'},
        {label: 'Type', flex: 2, sortKey: 'type'},
        {label: 'URI', flex: 2, sortKey: 'uri', },
        // TODO: Get timestamp from the event that created this artifact.
        // {label: 'Created at', flex: 1, sortKey: 'created_at'},
      ],
      rows: [],
      expandedRows: new Map(),
    };
    this.reload = this.reload.bind(this);
    this.toggleRowExpand = this.toggleRowExpand.bind(this);
    this.getExpandedArtifactsRow = this.getExpandedArtifactsRow.bind(this);
  }

  public getInitialToolbarState(): ToolbarProps {
    return {
      actions: [],
      breadcrumbs: [],
      pageTitle: 'Artifacts',
    };
  }

  public render(): JSX.Element {
    const {rows, columns} = this.state;
    return (
      <div className={classes(commonCss.page, padding(20, 'lr'))}>
        <CustomTable ref={this.tableRef}
          columns={columns}
          rows={rows}
          disablePaging={true}
          disableSelection={true}
          reload={this.reload}
          initialSortColumn='pipelineName'
          initialSortOrder='asc'
          getExpandComponent={this.getExpandedArtifactsRow}
          toggleExpansion={this.toggleRowExpand}
          emptyMessage='No artifacts found.' />
      </div>
    );
  }

  public async refresh(): Promise<void> {
    if (this.tableRef.current) {
      await this.tableRef.current.reload();
    }
  }

  private async reload(request: ListRequest): Promise<string> {
    // TODO: Consider making an Api method for returning and caching types
    if (!this.artifactTypes || !this.artifactTypes.size) {
      this.artifactTypes = await this.getArtifactTypes();
    }
    const {artifacts} = this.state;
    if (!artifacts.length) {
      try {
        const response = await this.api.metadataService.listArtifacts2();
        this.setState({artifacts: (response && response.artifacts) || []});
        this.clearBanner();
      } catch (err) {
        this.showPageError('Unable to retrieve Artifacts.', err);
      }
    }
    this.setState({
      rows: this.getRowsFromArtifacts(request),
    });
    return '';
  }

  private async getArtifactTypes(): Promise<Map<number, ArtifactType>> {
    const {error, response} =
        await this.api.metadataStoreService.getArtifactTypes(new GetArtifactTypesRequest());

    if (error) {
      this.showPageServiceError(
          'Unable to retrieve Artifact Types, some features may not work.', error);
      return new Map();
    }
    const artifactTypesMap = new Map<number, ArtifactType>();

    (response!.getArtifactTypesList() || []).forEach((artifactType) => {
      artifactTypesMap.set(artifactType.getId()!, artifactType);
    });

    return artifactTypesMap;
  }

  /**
   * Temporary solution to apply sorting, filtering, and pagination to the
   * local list of artifacts until server-side handling is available
   * TODO: Replace once https://github.com/kubeflow/metadata/issues/73 is done.
   * @param request
   */
  private getRowsFromArtifacts(request: ListRequest): Row[] {
    const collapsedAndExpandedRows = groupRows(this.state.artifacts
      .map((a) => { // Flattens
        const typeNumber = Number(a.type_id);
        const type = this.artifactTypes && this.artifactTypes.get(typeNumber!) ?
          this.artifactTypes.get(typeNumber!)!.getName() : typeNumber;
        return {
          id: `${type}:${a.id}`, // Join with colon so we can build the link
          otherFields: [
            getResourceProperty(a, ArtifactProperties.PIPELINE_NAME)
              || getResourceProperty(a, ArtifactCustomProperties.WORKSPACE, true),
            getResourceProperty(a, ArtifactProperties.NAME),
            a.id,
            type,
            a.uri,
            // TODO: Get timestamp from the event that created this artifact.
            // formatDateString(
            //   getArtifactProperty(a, ArtifactProperties.CREATE_TIME) || ''),
          ],
        } as Row;
      })
      .filter(rowFilterFn(request))
      .sort(rowCompareFn(request, this.state.columns)));

    this.setState({ expandedRows: collapsedAndExpandedRows.expandedRows });
    return collapsedAndExpandedRows.collapsedRows;
  }

  /**
   * Toggles the expansion state of a row
   * @param index
   */
  private toggleRowExpand(index: number): void {
    const {rows} = this.state;
    if (!rows[index]) {
      return;
    }
    rows[index].expandState = rows[index].expandState === ExpandState.EXPANDED ?
      ExpandState.COLLAPSED : ExpandState.EXPANDED;
    this.setState({rows});
  }

  private getExpandedArtifactsRow(index: number): React.ReactNode {
    return getExpandedRow(this.state.expandedRows, this.state.columns)(index);
  }
}

export default ArtifactList;