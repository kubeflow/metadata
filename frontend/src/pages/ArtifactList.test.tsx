import {
  Api,
  Artifact,
  ArtifactType,
  GetArtifactsResponse,
  GetArtifactTypesResponse,
} from '@kubeflow/frontend';
import * as React from 'react';
import ArtifactList from './ArtifactList';
import {PageProps} from './Page';
import {shallow, ShallowWrapper, ReactWrapper} from 'enzyme';
import TestUtils, {stringValue} from '../TestUtils';
import {RoutePage} from '../components/Router';
import CustomTable, {ExpandState} from '../components/CustomTable';

describe('ArtifactList', () => {
  let tree: ShallowWrapper | ReactWrapper;
  const updateBannerSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  const mockGetArtifactTypes = jest.spyOn(Api.getInstance().metadataStoreService, 'getArtifactTypes');
  const mockGetArtifacts = jest.spyOn(Api.getInstance().metadataStoreService, 'getArtifacts');

  const fakeGetArtifactTypesResponse = new GetArtifactTypesResponse();

  const artifactType1 = new ArtifactType();
  artifactType1.setId(1);
  artifactType1.setName('kubeflow.org/alpha/metrics');
  fakeGetArtifactTypesResponse.addArtifactTypes(artifactType1);

  const artifactType2 = new ArtifactType();
  artifactType2.setId(2);
  artifactType2.setName('kubeflow.org/alpha/data_set');
  fakeGetArtifactTypesResponse.addArtifactTypes(artifactType2);

  const artifactType3 = new ArtifactType();
  artifactType3.setId(3);
  artifactType3.setName('kubeflow.org/alpha/model');
  fakeGetArtifactTypesResponse.addArtifactTypes(artifactType3);

  const fakeGetArtifactsResponse = new GetArtifactsResponse();

  const artifact1 = new Artifact();
  artifact1.setId(1);
  artifact1.setTypeId(3);
  artifact1.setUri('gs://my-bucket/mnist');
  const artifact1PropertiesMap = artifact1.getPropertiesMap();
  artifact1PropertiesMap.set('name', stringValue('model'));
  artifact1PropertiesMap.set('pipeline_name', stringValue('pipeline-1'));
  artifact1PropertiesMap.set('version', stringValue('v0'));
  artifact1PropertiesMap.set('create_time', stringValue('2019-06-12T01:21:48.259263Z'));
  fakeGetArtifactsResponse.addArtifacts(artifact1);

  const artifact2 = new Artifact();
  artifact2.setId(2);
  artifact2.setTypeId(2);
  artifact2.setUri('gs://my-bucket/dataset2');
  const artifact2PropertiesMap = artifact2.getPropertiesMap();
  artifact2PropertiesMap.set('name', stringValue('dataset'));
  artifact2PropertiesMap.set('pipeline_name', stringValue('pipeline-1'));
  artifact2PropertiesMap.set('version', stringValue('v1'));
  artifact2PropertiesMap.set('create_time', stringValue('2019-06-16T01:21:48.259263Z'));
  const artifact2CustomPropertiesMap = artifact2.getCustomPropertiesMap();
  artifact2CustomPropertiesMap.set('__kf_workspace__', stringValue('workspace-1'));
  fakeGetArtifactsResponse.addArtifacts(artifact2);

  const artifact3 = new Artifact();
  artifact3.setId(3);
  artifact3.setTypeId(2);
  artifact3.setUri('gs://my-bucket/dataset2');
  const artifact3PropertiesMap = artifact3.getPropertiesMap();
  artifact3PropertiesMap.set('name', stringValue('dataset'));
  artifact3PropertiesMap.set('pipeline_name', stringValue('pipeline-1'));
  artifact3PropertiesMap.set('version', stringValue('v2'));
  artifact3PropertiesMap.set('create_time', stringValue('2019-07-01T01:00:00.000000Z'));
  const artifact3CustomPropertiesMap = artifact3.getCustomPropertiesMap();
  artifact3CustomPropertiesMap.set('__kf_workspace__', stringValue('workspace-1'));
  fakeGetArtifactsResponse.addArtifacts(artifact3);

  const artifact4 = new Artifact();
  artifact4.setId(4);
  artifact4.setTypeId(1);
  artifact4.setUri('gcs://my-bucket/mnist-eval.csv');
  const artifact4PropertiesMap = artifact4.getPropertiesMap();
  artifact4PropertiesMap.set('create_time', stringValue('2019-06-28T01:40:11.843625Z'));
  artifact4PropertiesMap.set('data_set_id', stringValue('13'));
  artifact4PropertiesMap.set(
      'description', stringValue('validating the MNIST model to recognize handwritten digits'));
  artifact4PropertiesMap.set('metrics_type', stringValue('validation'));
  artifact4PropertiesMap.set('model_id', stringValue('1'));
  artifact4PropertiesMap.set('name', stringValue('MNIST-evaluation'));
  artifact4PropertiesMap.set('owner', stringValue('someone@kubeflow.org'));
  const artifact4CustomPropertiesMap = artifact4.getCustomPropertiesMap();
  artifact4CustomPropertiesMap.set('__kf_run__', stringValue('run-2019-06-28T01:40:11.735816'));
  artifact4CustomPropertiesMap.set('__kf_workspace__', stringValue('ws1'));
  fakeGetArtifactsResponse.addArtifacts(artifact4);

  const artifact5 = new Artifact();
  artifact5.setId(5);
  artifact5.setTypeId(3);
  artifact5.setUri('gs://my-bucket/mnist');
  const artifact5PropertiesMap = artifact5.getPropertiesMap();
  artifact5PropertiesMap.set('name', stringValue('model'));
  artifact5PropertiesMap.set('pipeline_name', stringValue('pipeline-1'));
  artifact5PropertiesMap.set('version', stringValue('v1'));
  artifact5PropertiesMap.set('create_time', stringValue('2019-07-01T00:00:00.000000Z'));
  const artifact5CustomPropertiesMap = artifact5.getCustomPropertiesMap();
  artifact5CustomPropertiesMap.set('__kf_workspace__', stringValue('workspace-1'));
  fakeGetArtifactsResponse.addArtifacts(artifact5);

  function generateProps(): PageProps {
    return TestUtils.generatePageProps(
      ArtifactList,
      {pathname: RoutePage.ARTIFACTS} as any,
      '' as any,
      historyPushSpy,
      updateBannerSpy,
      jest.fn(),
      updateToolbarSpy,
      jest.fn());
  }

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => tree.unmount());

  it('Renders with no data', () => {
    tree = shallow(<ArtifactList {...generateProps()} />);

    expect(tree).toMatchSnapshot();
  });

  it('Renders with a list of Artifacts', async () => {
    mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
    mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
    tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

    await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
    await TestUtils.flushPromises();
    tree.update();
    expect(tree).toMatchSnapshot();
  });

  it('Renders and applies filter to a list of Artifacts', async () => {
    mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
    mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
    tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

    await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
    await TestUtils.flushPromises();
    tree.update();

    const table = tree.find(CustomTable).instance() as CustomTable;
    table.reload({filter: 'evaluation'});
    tree.update();

    expect(table.props.rows.length).toBe(1);
    expect(tree).toMatchSnapshot();
  });

  it('Renders and sorts Artifacts in ascending order by pipeline/workspace name',
    async () => {
      mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
      mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
      tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

      await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
      await TestUtils.flushPromises();
      tree.update();

      const table = tree.find(CustomTable).instance() as CustomTable;
      table.reload({sortBy: 'pipelineName', orderAscending: true});
      tree.update();

      const rows = table.props.rows.map((r) => r.otherFields[0]);
      expect(rows).toEqual(['pipeline-1', 'ws1']);
      expect(tree).toMatchSnapshot();
    });

  it('Renders and sorts Artifacts in ascending order by id',
    async () => {
      mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
      mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
      tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

      await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
      await TestUtils.flushPromises();
      tree.update();

      const table = tree.find(CustomTable).instance() as CustomTable;
      table.reload({sortBy: 'id', orderAscending: false});
      tree.update();

      const rows = table.props.rows.map((r) => r.otherFields[0]);
      expect(rows).toEqual(['pipeline-1', 'ws1']);
      expect(tree).toMatchSnapshot();
    });

  it('Renders and expands artifacts from pipeline-1',
    async () => {
      mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
      mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
      tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

      await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
      await TestUtils.flushPromises();
      tree.update();

      const table = tree.find(CustomTable).instance() as CustomTable;
      const index = table.props.rows.findIndex((r) => r.otherFields[0] === 'pipeline-1');
      tree.find('IconButton.expandButton').at(index).simulate('click');
      expect(table.props.rows[index].expandState).toBe(ExpandState.EXPANDED);
      const expandedRows = tree.find('.expandedContainer CustomTableRow');
      expect(expandedRows.length).toBe(4);
      expect(expandedRows.get(0).props.row.id).toBe('kubeflow.org/alpha/model:1');
      expect(expandedRows.get(1).props.row.id).toBe('kubeflow.org/alpha/data_set:2');

      tree.find('IconButton.expandButton').at(index).simulate('click');
      expect(table.props.rows[index].expandState).toBe(ExpandState.COLLAPSED);
      expect(tree.find('.expandedContainer').exists()).toBeFalsy();
    });

  it('Renders Artifact with no grouped rows with placeholder',
    async () => {
      mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
      mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
      tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

      await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
      await TestUtils.flushPromises();
      tree.update();

      const table = tree.find(CustomTable).instance() as CustomTable;
      const index = table.props.rows.findIndex((r) => r.otherFields[0] === 'ws1');
      expect(table.props.rows[index].expandState).toBe(ExpandState.NONE);
      expect(tree.find('.expandButtonPlaceholder').exists())
      expect(tree.find('.expandedContainer').exists()).toBeFalsy();
      expect(tree).toMatchSnapshot();
    });

  it('Shows error when artifacts cannot be retrieved', async () => {
    mockGetArtifactTypes.mockResolvedValue(fakeGetArtifactTypesResponse);
    // @ts-ignore
    mockGetArtifacts.mockResolvedValue();
    tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

    await Promise.all([mockGetArtifactTypes, mockGetArtifacts]);
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledWith(expect.objectContaining({
      message: 'Unable to retrieve Artifacts.',
      mode: 'error',
    }));
  });

  it('Shows error when artifact types cannot be retrieved', async () => {
    mockGetArtifacts.mockResolvedValue(fakeGetArtifactsResponse);
    // @ts-ignore
    mockGetArtifactTypes.mockResolvedValue();
    tree = TestUtils.mountWithRouter(<ArtifactList {...generateProps()} />);

    await Promise.all([mockGetArtifactTypes, mockGetArtifactTypes]);
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledWith(expect.objectContaining({
      message:
        'Unable to retrieve Artifact Types, some features may not work.',
      mode: 'error',
    }));
  });
});
