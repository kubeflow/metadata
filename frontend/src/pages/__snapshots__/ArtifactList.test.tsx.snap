// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtifactList Renders with a list of Artifacts 1`] = `
<ArtifactList
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "pathname": "/artifacts",
    }
  }
  match=""
  toolbarProps={
    Object {
      "actions": Array [],
      "breadcrumbs": Array [],
      "pageTitle": "Artifacts",
    }
  }
  updateBanner={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {},
        ],
        Array [
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  updateDialog={[MockFunction]}
  updateSnackbar={[MockFunction]}
  updateToolbar={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "actions": Array [],
            "breadcrumbs": Array [],
            "pageTitle": "Artifacts",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="page_f1flacxk f9qqjqd"
  >
    <CustomTable
      columns={
        Array [
          Object {
            "customRenderer": [Function],
            "flex": 1,
            "label": "Name",
          },
          Object {
            "flex": 1,
            "label": "Version",
          },
          Object {
            "flex": 1,
            "label": "Type",
          },
          Object {
            "flex": 1,
            "label": "URI",
          },
          Object {
            "flex": 1,
            "label": "Workspace",
          },
          Object {
            "flex": 1,
            "label": "Created at",
          },
        ]
      }
      disablePaging={true}
      disableSelection={true}
      emptyMessage="No artifacts found."
      reload={[Function]}
      rows={
        Array [
          Object {
            "id": "1",
            "otherFields": Array [
              "model",
              "v1",
              "1",
              "gs://my-bucket/mnist",
              "workspace-1",
              "6/11/2019, 6:21:48 PM",
            ],
          },
          Object {
            "id": "2",
            "otherFields": Array [
              "dataset",
              "v2",
              "3",
              "gs://my-bucket/dataset2",
              "workspace-1",
              "6/15/2019, 6:21:48 PM",
            ],
          },
        ]
      }
    >
      <div
        className="pageOverflowHidden_f15djeb2"
      >
        <div
          className="header_f15jqnss feotlki"
        >
          <div
            className="columnName_fcye2mf"
            key="0"
            style={
              Object {
                "width": "16.666666666666664%",
              }
            }
          >
            <div>
              Name
            </div>
          </div>
          <div
            className="columnName_fcye2mf"
            key="1"
            style={
              Object {
                "width": "16.666666666666664%",
              }
            }
          >
            <div>
              Version
            </div>
          </div>
          <div
            className="columnName_fcye2mf"
            key="2"
            style={
              Object {
                "width": "16.666666666666664%",
              }
            }
          >
            <div>
              Type
            </div>
          </div>
          <div
            className="columnName_fcye2mf"
            key="3"
            style={
              Object {
                "width": "16.666666666666664%",
              }
            }
          >
            <div>
              URI
            </div>
          </div>
          <div
            className="columnName_fcye2mf"
            key="4"
            style={
              Object {
                "width": "16.666666666666664%",
              }
            }
          >
            <div>
              Workspace
            </div>
          </div>
          <div
            className="columnName_fcye2mf"
            key="5"
            style={
              Object {
                "width": "16.666666666666664%",
              }
            }
          >
            <div>
              Created at
            </div>
          </div>
        </div>
        <div
          className="scrollContainer_fea2b5v"
          style={
            Object {
              "minHeight": 60,
            }
          }
        >
          <div
            className="expandableContainer_f1l0zr01"
            key="0"
          >
            <div
              className="tableRow row_f1xdiyip feotlki"
              onClick={[Function]}
              role="checkbox"
              tabIndex={-1}
            >
              <div
                className="cell_fy8nr2z"
                key="0"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                <Link
                  className="link_f1d44nl9"
                  onClick={[Function]}
                  replace={false}
                  to="/artifacts/models/1"
                >
                  <a
                    className="link_f1d44nl9"
                    href="/artifacts/models/1"
                    onClick={[Function]}
                  >
                    model
                  </a>
                </Link>
              </div>
              <div
                className="cell_fy8nr2z"
                key="1"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                v1
              </div>
              <div
                className="cell_fy8nr2z"
                key="2"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                1
              </div>
              <div
                className="cell_fy8nr2z"
                key="3"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                gs://my-bucket/mnist
              </div>
              <div
                className="cell_fy8nr2z"
                key="4"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                workspace-1
              </div>
              <div
                className="cell_fy8nr2z"
                key="5"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                6/11/2019, 6:21:48 PM
              </div>
            </div>
          </div>
          <div
            className="expandableContainer_f1l0zr01"
            key="1"
          >
            <div
              className="tableRow row_f1xdiyip feotlki"
              onClick={[Function]}
              role="checkbox"
              tabIndex={-1}
            >
              <div
                className="cell_fy8nr2z"
                key="0"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                <Link
                  className="link_f1d44nl9"
                  onClick={[Function]}
                  replace={false}
                  to="/artifacts/models/2"
                >
                  <a
                    className="link_f1d44nl9"
                    href="/artifacts/models/2"
                    onClick={[Function]}
                  >
                    dataset
                  </a>
                </Link>
              </div>
              <div
                className="cell_fy8nr2z"
                key="1"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                v2
              </div>
              <div
                className="cell_fy8nr2z"
                key="2"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                3
              </div>
              <div
                className="cell_fy8nr2z"
                key="3"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                gs://my-bucket/dataset2
              </div>
              <div
                className="cell_fy8nr2z"
                key="4"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                workspace-1
              </div>
              <div
                className="cell_fy8nr2z"
                key="5"
                style={
                  Object {
                    "width": "16.666666666666664%",
                  }
                }
              >
                6/15/2019, 6:21:48 PM
              </div>
            </div>
          </div>
        </div>
      </div>
    </CustomTable>
  </div>
</ArtifactList>
`;

exports[`ArtifactList Renders with no data 1`] = `
<div
  className="page_f1flacxk f9qqjqd"
>
  <CustomTable
    columns={
      Array [
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Name",
        },
        Object {
          "flex": 1,
          "label": "Version",
        },
        Object {
          "flex": 1,
          "label": "Type",
        },
        Object {
          "flex": 1,
          "label": "URI",
        },
        Object {
          "flex": 1,
          "label": "Workspace",
        },
        Object {
          "flex": 1,
          "label": "Created at",
        },
      ]
    }
    disablePaging={true}
    disableSelection={true}
    emptyMessage="No artifacts found."
    reload={[Function]}
    rows={Array []}
  />
</div>
`;
